ouroboros-consensus-shelley-test/test/Test/ThreadNet/RealTPraos.hs

```diff
@@ -56,36 +68,88 @@ data TestSetup = TestSetup
 
 instance Arbitrary TestSetup where
   arbitrary = do
-    setupD <- (/10)         <$> choose   (0, 10)
-    setupK <- SecurityParam <$> elements [5, 10]
+      setupD  <- arbitrary
+      setupD2 <- arbitrary
+      setupK  <- SecurityParam <$> choose (minK, maxK)
 
-    setupTestConfig <- arbitrary
+      setupTestConfig <- genRealTPraosTestConfig setupK
 
-    setupVersion   <- genVersion (Proxy @(ShelleyBlock Crypto))
+      setupVersion <- genVersion (Proxy @(ShelleyBlock Crypto))
 
-    pure TestSetup
-      { setupD
-      , setupK
-      , setupTestConfig
-      , setupVersion
-      }
+      pure TestSetup
+        { setupD
+        , setupD2
+        , setupK
+        , setupTestConfig
+        , setupVersion
+        }
 
   -- TODO shrink
 
+minK :: Word64
+minK = 5   -- Less than this increases risk of CP violations
+
+maxK :: Word64
+maxK = 10   -- More than this wastes execution time
+
+-- | In particular, choose a 'numSlots' appropriate for this test
+genRealTPraosTestConfig :: SecurityParam -> Gen TestConfig
+genRealTPraosTestConfig setupK = do
+    tc <- arbitrary
+
+    ensureMultipleEpoches <- frequency [(2, pure True), (1, pure False)]
+    pure $ if not ensureMultipleEpoches then tc else
+      let mkEpochSize :: SecurityParam -> Word64
+          mkEpochSize k = 10 * maxRollbacks k * recipActiveSlotsCoeff
```

Comment:
We have this in `mkGenesisConfig` too, maybe promote it to a top-level helper?

Reply:
Done, thanks!